src/clean-core/to_string.cc

```diff
@@ -473,265 +473,159 @@ char const* unsigned_int_to_binary(IntType value, char* end)
 template <class IntType>
 void unsigned_to_string_impl(cc::string_stream_ref ss, IntType value, parsed_fmt_args const& args)
 {
+    auto const add_fill = [&](int length) {
+        auto const fill = args.sign_aware_zero_padding ? '0' : args.fill;
+        if (args.width > 0 && length < args.width)
+            for (auto i = 0; i < args.width - length; ++i)
+                ss << fill;
+    };
```

Comment:
not sure if it works here precisely, but printf has builtin support for zero paddings, for example
`printf("%06d\n", 123);` outputs `000123` (zero-padding to 6 digits)

Reply:
It does, however I decided against modifying the printf format string programatically and leave it as a char literal as that introduces warnings. On the other hand, IF we start modifying the printf format string much of the formatting code can be simplified.

Reply:
There's no need to modify the format string for that:
`printf("%0*d\n", 8, 123);` outputs 123 zero-padded to 8 digits, a dynamic value.

Reply:
Good point, I forgot you can do that :). I'll change that!

Reply:
On second thought, I believe I would still have to modify the format string in this case to assure the correct padding is used. 
`printf("%0*d", 8, 123)` is different from `printf("%*d", 8, 123)`.

Reply:
isn't that mode already being branched on?

Reply:
Just so we're talking about the same thing: You suggest we remove `add_fill` by replacing it with `printf("%0*d", 8 , 123)`. If `args.sign_aware_zero_padding == true` we expect to get `00000123`, otherwise `      123`. Correct?

Reply:
Yeah, im thinking you could trim it down to something like
```C++
if (pad_with_zeroes)
    snprintf("%0*llu", width, uint64_t(value));
else
    snprintf("%*llu", width, uint64_t(value));

Reply:
`args.fill` can be *almost* any character. Although cppref seems to suggest that it is possible to use other chars than space or `0` ("The result is padded with space characters (by default)") , I can't figure out how to achieve this behavior.

Reply:
Oh so you can pad with any char? Im not sure if we should go out of our way to support that case, but if yes, you could do something like
```C++
int printed_width = snprintf("%llu", val);
char fill[] = "XXXXXXXXXXXXXXXXXXXXXXXXX"; // (duplicate fill char to the buffer)
snprintf("%*s", width - printed_width, fill);
```

Reply:
Yeah, we can pad with anything that is not `{` or `}`. Therefore I think `add_fill` is a decent solution.

Reply:
I see, yep